src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -4579,6 +4618,85 @@ namespace winrt::TerminalApp::implementation
         _updateThemeColors();
     }
 
+    winrt::fire_and_forget TerminalPage::_ControlMenuChangedHandler(const IInspectable /*sender*/,
+                                                                    const MenuChangedEventArgs args)
+    {
+        if constexpr (!Feature_ShellCompletions::IsEnabled())
+        {
+            co_return;
+        }
+
+        // User must explicitly opt-in on Preview builds
+        if (!_settings.GlobalSettings().EnableShellCompletionMenu())
+        {
+            co_return;
+        }
+
+        auto weakThis{ get_weak() };
+        co_await winrt::resume_background();
+        if (const auto& page{ weakThis.get() })
+        {
+            // `this` is safe to use
+
+            // Parse the json
+            try
+            {
+                auto commandsCollection = Command::ParsePowerShellMenuComplete(args.MenuJson(),
+                                                                               args.ReplacementLength());
```

Comment:
Hmm. That's a fair point, but maybe the resume should be elsewhere. 

Right now, this gets triggered by the output thread, so handling it happens on the thread draining the output. 

This should probably be moved, so that the `resume_background` happens in the control itself, so it can immediately bail and unblock the output thread.